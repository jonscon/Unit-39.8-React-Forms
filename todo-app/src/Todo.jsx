import { useState } from "react";

function Todo({ id = "1", task = "default todo", update, remove }) {
    const [editTask, setEditTask] = useState(task);
    const [isEditing, setIsEditing] = useState(false);

    const toggleEdit = () => {
        setIsEditing(edit => !edit);
    }

    const handleChange = e => {
        setEditTask(e.target.value);
    }

    const handleRemove = () => remove(id);

    const handleUpdate = e => {
        e.preventDefault();
        update(id, editTask);
        setIsEditing(false);
    }

    // default todo view
    let jsx = (
        <div>
            <li>{ task }</li>
            <button onClick={ toggleEdit }>Edit</button>
            <button onClick={ handleRemove }>X</button>
        </div>
    );

    // change todo view if editing
    if (isEditing) {
        jsx = (
            <div>
                <form onSubmit={ handleUpdate }>
                    <input value={ editTask } onChange={ handleChange }/>
                    <button>Update</button>
                </form>
            </div>
        )
    }
    return jsx;
}

export default Todo;