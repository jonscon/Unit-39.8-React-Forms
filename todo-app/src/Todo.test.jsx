import { render, fireEvent } from "@testing-library/react";
import Todo from "./Todo.jsx";
import {it, expect, vi } from 'vitest';


// Smoke Test
it("renders without crashing", function() {
    render(<Todo />);
});

// Snapshot Test
it("matches snapshot", function() {
    const {asFragment} = render(<Todo />);
    expect(asFragment()).toMatchSnapshot();
});

it("runs the update function on submit", function() {
    const updateMock = vi.fn();
    const { getByText } = render(<Todo update={updateMock} />);
    const editButton = getByText("Edit");
    fireEvent.click(editButton);
    const updateButton = getByText("Update");
    fireEvent.click(updateButton);
    expect(updateMock).toHaveBeenCalled();
})

it("runs the remove function on submit", function() {
    const removeMock = vi.fn();
    const { getByText } = render(<Todo remove={removeMock} />);
    const removeButton = getByText("X");
    fireEvent.click(removeButton);
    expect(removeMock).toHaveBeenCalled();
})