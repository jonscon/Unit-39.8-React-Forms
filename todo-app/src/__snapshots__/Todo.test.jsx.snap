// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <li>
      default todo
    </li>
    <button>
      Edit
    </button>
    <button>
      X
    </button>
  </div>
</DocumentFragment>
`;
